src/python/WMCore/WMSpec/StdSpecs/StepChain.py

```diff
@@ -348,9 +348,27 @@ def setupNextSteps(self, task, origArgs):
             currentCmsswStepHelper.keepOutput(childKeepOutput)
             self.setupOutputModules(task, taskConf, currentCmsRun, childKeepOutput)
 
+            # Get and append campaign name for each step
+            if taskConf.get("Campaign", None) is not None:
+                campaignNames.append(taskConf.get("Campaign"))
+            else:
+                # check if workload has a campaign name
+                if self.workload.getCampaign() is not None:
+                    campaignNames.append(self.workload.getCampaign())
+
         # Closing out the task configuration. The last step output must be saved/merged
         currentCmsswStepHelper.keepOutput(True)
 
+        # Set campaign name
+        if campaignNames:
+            # Add first task campaign name if it exists
+            if task.getCampaignName() is not None:
+                campaignNames = [task.getCampaignName()] + campaignNames
+            # Report only unique values
+            seen = set()
+            uniqueCampaignNames = [x for x in campaignNames if not (x in seen or seen.add(x))]
```

Comment:
It might be worth it adding a comment line saying that it is unique and it keeps the original content order.